### PR TITLE
cpp_wrapper: Complete support for Layout arguments

### PR DESCRIPTION
**Pull request by benjaminglass1**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #141699
* #141371
* #141370
* #141369
* #141580
* __->__ #141368

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

File: torch/_inductor/codegen/cpp_wrapper_cpu.py [hunk 1]
```diff
@@ -1874,6 +1874,13 @@ def generate_py_arg_inner(lines, raw_arg, arg_type):
                     self.include_extra_header("torch/csrc/utils/pythoncapi_compat.h")
                 self.include_extra_header("torch/csrc/DynamicTypes.h")
                 return f"Py_NewRef(torch::getTHPDtype(static_cast<c10::ScalarType>({self.codegen_dtype(raw_arg)})))"
+            elif isinstance(raw_arg, torch.layout):
+                # dtype
```
Comment (benjaminglass1) [inline torch/_inductor/codegen/cpp_wrapper_cpu.py:1878]:
```suggestion
                # memory layout
```

**Review by desertfire — APPROVED** [review]

File: torch/_inductor/codegen/cpp_wrapper_cpu.py [hunk 2]
```diff
@@ -1879,6 +1879,13 @@ def generate_py_arg_inner(lines, raw_arg, arg_type):
                     self.include_extra_header("torch/csrc/utils/pythoncapi_compat.h")
                 self.include_extra_header("torch/csrc/DynamicTypes.h")
                 return f"Py_NewRef(torch::getTHPDtype(static_cast<c10::ScalarType>({self.codegen_dtype(raw_arg)})))"
+            elif isinstance(raw_arg, torch.layout):
+                # memory layout
+                if sys.version_info < (3, 10):
```
[the remark this answers is not in the document]
Comment (benjaminglass1) [reply torch/_inductor/codegen/cpp_wrapper_cpu.py:1884]: Will do.

**Comment by benjaminglass1** [comment]
@desertfire It appears that this PR netted us some performance gains I didn't notice! I've updated with the new performance numbers from the CI error message, and I'll get this landed.

**Pull request closed without merging**